documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md

```diff
@@ -0,0 +1,586 @@
+ingestion server as well.
+
+Finally we will update the monorepo's root
+[`docker-compose.yml`](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/docker-compose.yml)
+to add a new `catalog-indexer-worker` service. Its build context should point to
+the nested `data_refresh/indexer_worker` directory, and it should map the
+exposed port to enable the API to be reached by the catalog.
+
+When this work is complete, it should be possible to run `just catalog/ipython`
+and curl the new indexer worker. The existing ingestion-server and
+indexer-worker services are unaffected (it is still possible to run legacy data
+refreshes locally and in production).
+
+### Implement distributed reindexing locally
+
+In this step we will add tasks to the data refresh DAGs to orchestrate the
+distributed reindex. At the end of this step, it will be possible to run a
+distributed reindex locally. Because the infrastructure work to create the EC2
+instances is not complete, it can not be run on production yet. The following
+code can all be refactored from
+[`distributed_reindex_scheduler.py`](https://github.com/WordPress/openverse/blob/main/ingestion_server/ingestion_server/distributed_reindex_scheduler.py).
+
+- Use Airflow's
+  [EC2Hook#describe_instances](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/hooks/ec2/index.html#airflow.providers.amazon.aws.hooks.ec2.EC2Hook.describe_instances)
+  to get a list of internal URLs bound to the indexer workers for the
+  appropriate environment. In the local environment, this will instead return
+  the URL for the catalog-indexer-worker Docker container.
+- Use
+  [EC2Hook#start_instances](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/hooks/ec2/index.html#airflow.providers.amazon.aws.hooks.ec2.EC2Hook.start_instances)
+  to start the appropriate EC2 instances. This step should skip in a local
+  environment.
+- Use dynamic task mapping to generate a Sensor for each of the expected indexer
+  workers, and ping its `healthcheck` endpoint until it passes.
+- Use dynamic task mapping to distribute reindexing across the indexer workers
+  by first calculating `start` and `end` indices that will split the records in
+  the media table into even portions, depending on the number of workers
+  available in the given environment. Then:
+  - POST to each worker's `reindexing_task` endpoint the `start_index` and
+    `end_index` it should handle
+  - Use a Sensor to ping the worker's `task/{task_id}` endpoint until the task
+    is complete, logging the progress as it goes
+  - Use the
+    [EC2Hook#stop_instances](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/hooks/ec2/index.html#airflow.providers.amazon.aws.hooks.ec2.EC2Hook.stop_instances)
+    to shut the instance down. This step should skip in a local environment.
```

Comment:
You mention using `ec2-service` later on, and in thinking further about the implications of that compared to creating EC2 instances directly, I think we have an opportunity to leverage the functionality of an auto-scaling group to trivially get somewhat automatically updating indexer workers, along with simplifying the DAGs set of tasks by removing most of the need for mapped tasks.

When managing EC2 instances directly (what we do currently), we start the instances once when we provision them, and then immediately shut them down. When the instances start again, a cron job is responsible for restarting the docker compose services[^1]. All of this is configured once in the user data script at instance creation. User data scripts run only on instance creation and never again (at least not automatically, I'm sure you could locate the script and run it manually if you wanted to).

[^1]: Which, by the way, really is not necessary if we just used the docker service as intended, with an appropriate restart policy, but that's neither here nor there

With ASGs, however, we have some interesting capabilities that would make deploying even a bumped version tag (if we did not use a new tag to point to the "release" we want). The critical thing to know about ASGs is that they do not rely on _stopped_ instances, but rather _terminated_ instances. Meaning the instances literally get created and destroyed as needed, rather than started and stopped. This is important because it means the user data script runs every time an ASG instance spins up. The user data script is defined in the launch template, which in turn means that a "deployment" for a correctly leveraged ASG should mean just updating the launch template and then replacing each running instance one-by-one (or in whatever pattern makes sense for the given service). If we ran the API or frontend in an ASG again, that's how we would do it, rather than needing to create an entirely new ASG each time.

For the indexer workers, what does this mean? It means we can actually basically ignore Ansible for them, and instead completely leverage Terraform and only Terraform, significantly simplifying their management. Rather than running an Ansible playbook that would one-by-one start and then SSH into each indexer worker, we would just push an update to the ASG launch template. The next time any instance starts in that ASG, it will use the new launch template, and therefore have whatever updates we pushed to it.

This somewhat blends our old approach with the new approach, and basically sets up the best part of ECS (the need to only define a task-definition) with the ASG (by using a launch template, instead of task definition).

Then, in Airflow, rather than starting and stopping individual EC2 instances, we just do the following:

1. Query for the relevant ASG for the environment we need workers for
2. Issue a `set-desired-capacity` command to the ASG to set it to the capacity used for the data refresh
3. Query the ASG using `describe-auto-scaling-group`, which returns the ASG health and a list of all the instances with health status, and wait until instances are available to accept tasks (some or all are healthy).
4. Continue with the non-ASG approach as described in the IP, just use the ASG's instances list instead of querying for the EC2 instances matching the tag (and of course skip starting the instances).
5. As each instance finishes everything the DAG needs it to do, tell the ASG to spin it down using [manual specific instance termination](https://docs.aws.amazon.com/autoscaling/ec2/userguide/ec2-auto-scaling-scaling-manually.html#terminate-an-instance-aws-cli) which also allows us to decrement the desired capacity at the same time. This bascially tells the ASG: terminate this instance, and don't replace it.

This has some significant benefits over the EC2 approach:

1. Deployments will take literally seconds (updating a launch template is nearly instant)
2. Instances do not need to know how to terminate themselves, the DAG will do that for us
3. We can take a completely "hands-free" approach to the instances, thinking only in terms of launch templates and ASGs for the most part, eliminating the need for any new Ansible playbooks.
4. Additionally, we can use SSM parameters for selecting the AMI at launch time rather than using a static AMI ID, meaning we have one less place to worry about updating system dependencies (which is the main headache long-running EC2 instances carry, in my opinion).

In the future, we could even employ a target group with the internal load balancer and use target group stickiness configured with a custom app cookie, where each task responsible for one of the indexer workers generates a unique cookie value for the lifetime of the DAG run. Then the target group will automatically route requests to the same instance without the task needing to maintain a reference to the instance or us ever needing to even list the individual instances. That's definitely out of scope here, but wanted to share an example of what ASG's get us.

To summarise, rather than directly interacting with EC2, we can interact with the ASG. I don't think we can use AWS operators for this, there don't appear to be ones for ASGs, but boto3 supports them, so we just need to write our own interactions (like we do for s3). This eliminates the need for the Ansible playbook, like I said, so it will meaningfully reduce the amount of work done on the infrastructure side, in addition to reducing the number of mapped tasks in the DAG.

Reply:
Oh also, I forgot about this difference with ECS. We already have a process for automatically updating the ECS task definition when we cut a new release. We do not have that for launch templates, which are roughly the equivalent for EC2 instances (and the ASG approach). But, as you've said in the plan, we do not often update the code for the workers, so that isn't a big concern. Additionally, it would be good to keep updates to the launch templates manual (e.g., running `tf apply`) so that we make sure that we only do it when a data refresh is _not_ running.

Reply:
@sarayourfriend this is incredible 😍 I'm going to take some time to do a bit more background reading on this myself, but this is really exciting.

One possible correction is that while the indexer workers currently shut themselves down, in my proposed implementation the _DAG_ would send the `stop_instances` request. (However as you've noted that's only in the context of a data refresh; they need to shut themselves down after deploying. I also had not realized that we're currently using a cron job to restart the stack when the instances start 😮)

Reply:
+1-ing Staci's note - thank you for this @sarayourfriend! I hadn't thought about leveraging ASGs here for this, but it feels so _right_. Definitely on board with this approach!

Reply:
Oh, thanks for the clarification there, Staci.

> they need to shut themselves down after deploying

That would no longer be necessary using the ASG, just to clarify :grin:

Reply:
Wanted to clarify this as well:

> so that we make sure that we only do it [launch template updates] when a data refresh is not running

Updating the launch template does not necessarily automatically trigger a replacement of running instances that are using the old launch template. You _can_ configure ASGs to do that, but I'm pretty sure the default configuration does not automatically deploy the latest version of a launch template. The only potential issue is if for some reason one of the instances crashes (stops responding on healthchecks) and gets recreated, the ASG would use the latest launch template to do that, and if there's any potential for different versions of the indexer worker running at once to cause issues with completing the refresh, it'd be good to avoid that.

If that won't cause issues, then never mind, doesn't really matter, we can update the launch templates whenever we want.

Reply:
@sarayourfriend perhaps to your point above, during a data refresh we could set the ASG to use a fixed version of the template rather than `latest`, and then switch the config back to `latest` after the refresh is complete. That should be possible with boto3.

Reply:
> during a data refresh we could set the ASG to use a fixed version of the template rather than latest, and then switch the config back to latest after the refresh is complete.

Pinning the versions in Airflow is a really interesting idea 🤔 I think we could do something like this to pin/unpin the launch template version, but if the user_data script is always pulling the `latest` tag on the Docker image then code changes will be able to come through (for new instances) regardless, right? Although maybe those changes are okay to make mid-refresh.

One thing to note is that if an indexer worker fails for any reason, the Airflow `wait_for_reindexing` task will fail and won't automatically retry with a new worker instance. If the worker failed due to some issue with the reindexing, Airflow will terminate the instance to ensure it’s not left running (and prevent the ASG from making a new one). If the instance itself _crashes_ for some reason and the ASG tears it down, then the Airflow `wait_for_reindexing` task will still fail because it will try to poll a terminated instance. But the ASG will spin up a new one and the Airflow task will not know how to connect to resume, or to terminate the new instance gracefully. I think we’d need an extra Airflow task at the end to account for this 😓 But in either case, if a worker fails the DAG won’t be able to retry on a new instance automatically.

This discussion has made me realize that if we use an ASG instead of the more naive EC2 approach, we lose the ability to retry a single failed reindexer worker, because when a worker is terminated and a new one is spun up, there is no way to clear a single `trigger_reindex` task (because it will be trying to connect to the terminated instance). The best thing I have come up with is making `distributed_reindex` a reusable TaskGroup that accepts a list of worker params, and then having an entirely separate DAG that accepts worker params as a conf option and can be used to manually retry parts of a failed data refresh. This feels very convoluted.

Reply:
Although as we pointed out recently, if a single reindexing task fails and we're forced to retry _all of them_ (in this case by clearing the DAG at the task that sets the ASG desired capacity), it's not any _slower_ than retrying a single indexer worker because they run concurrently. The loss there is in the literal cost of unnecessarily running the other workers for however many hours the reindexing takes.

Reply:
Sorry to post back-to-back 😅 but -- the inability to retry a single reindexing task has an advantage in that it actually fixes the problem of ensuring that all the indexer workers are running the exact same code and have the exact same configuration. The trade-off is just the cost of running the successful workers in scenarios where the configuration _hadn't_ changed, which is perhaps fair.

This would allow us to use `latest` for both the launch template version and docker tag.

Reply:
We should generate AMIs for the indexer workers with the Docker image they should run bundled into them: https://github.com/WordPress/openverse-infrastructure/issues/843#issuecomment-2058094706

The launch template would always point to the latest AMI, and a new AMI would only be generated for production manually (we can also make it manual for staging if we want). We can configure the ASG to replace instances using the launch template they were running before, rather than using the latest launch template version (or pin the ASG to a specific launch template version rather than `$Latest`, and manually bump it when data refresh isn't running for that environment if a new AMI exists, maybe even just using a DAG that runs and checks if data refresh is running before bumping the launch template to the new AMI version).